olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp

```diff
@@ -126,6 +129,78 @@ client::CancellationToken ApiClientLookup::LookupApiClient(
       });
 }
 
+ApiClientLookup::ApiClientResponse ApiClientLookup::LookupApiSync(
+    const client::HRN& catalog,
+    client::CancellationContext cancellation_context, std::string service,
+    std::string service_version, FetchOptions options,
+    client::OlpClientSettings settings) {
+  repository::ApiCacheRepository repository(catalog, settings.cache);
+
+  if (options != OnlineOnly) {
+    auto url = repository.Get(service, service_version);
+    if (url) {
+      OLP_SDK_LOG_INFO_F(kLogTag, "getApiClient(%s, %s) -> from cache",
+                         service.c_str(), service_version.c_str());
+      client::OlpClient client;
+      client.SetSettings(std::move(settings));
+      client.SetBaseUrl(*url);
+      return std::move(client);
+    } else if (options == CacheOnly) {
+      return client::ApiError(
+          client::ErrorCode::NotFound,
+          "Cache only resource not found in cache (loopup api).");
+    }
+  }
+
+  auto client = std::make_shared<client::OlpClient>();
+  client->SetSettings(std::move(settings));
+
+  client::Condition condition(cancellation_context);
+
+  // when the network operation took too much time we cancel it and exit
+  // execution, to make sure that network callback will not access dangling
+  // references we protect them with atomic bool flag.
+  auto interest_flag = std::make_shared<std::atomic_bool>(true);
+
+  ApiClientLookup::ApiClientResponse api_response;
+  cancellation_context.ExecuteOrCancelled([&]() {
+    return ApiClientLookup::LookupApiClient(
+        client, service, service_version, catalog,
+        [&,
+         interest_flag](ApiClientLookup::ApiClientResponse response) mutable {
+          if (interest_flag->exchange(false)) {
+            api_response = std::move(response);
+            condition.Notify();
+          }
+        });
+  });
+
+  if (!condition.Wait()) {
+    // We are just about exit the execution.
+    interest_flag->store(false);
+
+    if (cancellation_context.IsCancelled()) {
+      // We can't use api response here because it could potentially be
+      // uninitialized.
+      return client::ApiError(client::ErrorCode::Cancelled,
+                              "Operation cancelled.");
+    } else {
+      cancellation_context.CancelOperation();
+      return client::ApiError(client::ErrorCode::RequestTimeout,
+                              "Network request timed out.");
+    }
+  }
+
+  if (api_response.IsSuccessful() && options != OnlineOnly) {
```

Comment:
Let's keep the logic as it is designed in other repositories
if (api_response.IsSuccessful()) {